### PR TITLE
Allow @differentiable(reverse) to be specified without primal or adjoint

### PR DESCRIPTION
**Pull request by rxwei**
In the past, the `@differentiable` attribute has only been used for primitive adjoint registration. In the future, `@differentiable` should be able to trigger AD when `adjoint:` is not provided, as something that guarantees a function to be differentiable.

The SIL `[reverse_differentiable]` attribute already triggers AD when there's no primal or adjoint provided. So this is merely a change to parsing and attribute checking.

Here's the new syntax for `@differentiable`
```
differentiable-attr ::= '@differentiable' '(' differentiable-attr-arguments ')'
differentiable-attr-arguments ::=
    ('forward' | 'reverse')
    (',' differentiable-attr-parameters)?
    (',' 'primal' ':' differentiable-attr-func-specifier)?
    (',' 'adjoint' ':' differentiable-attr-func-specifier)?
    generic-where-clause?
```

### CODE REVIEW

File: lib/SILGen/SILGen.cpp [hunk 1]
```diff
@@ -799,25 +799,25 @@ void SILGenModule::emitAbstractFuncDecl(AbstractFunctionDecl *AFD) {
...
-        getFunction(SILDeclRef(adjointFn), ForDefinition)->getName();
+      if (auto *adjointFn = diffAttr->getAdjointFunction()) {
+        // If the adjoint is specified but the primal is not, then we treat the
+        // original as the primal.
+        if (primName.empty())
+          primName = silOriginalFn->getName();
+        adjName = getFunction(SILDeclRef(adjointFn), ForDefinition)->getName();
+      }
+      else if (!primName.empty()) {
+        llvm_unreachable("Primal cannot be present if adjoint is not");
```
Comment (marcrasi) [inline lib/SILGen/SILGen.cpp]: I couldn't find logic that prevents this from happening, or a test that exercises the case where a user writes `@differentiable` with primal without adjoint. If the logic and/or the test are indeed not there, please add them.

**Review by marcrasi — CHANGES_REQUESTED** [review]

File: lib/SILGen/SILGen.cpp (hunk 1 above)
Comment (rxwei) [reply lib/SILGen/SILGen.cpp]: Done.

File: include/swift/AST/DiagnosticsParse.def [hunk 2]
```diff
@@ -1451,6 +1451,10 @@ ERROR(attr_differentiable_missing_label,PointsToFirstBadToken,
 ERROR(attr_differentiable_expected_colon_after_label,PointsToFirstBadToken,
       "expected a colon ':' after '%0'", (StringRef))
 
+ERROR(attr_differentiable_expected_config_after_comma,none,
+      "expected a configuration, e.g. 'withRespectTo:', 'primal:' or "
```
[the remark this answers is not in the document]
Comment (rxwei) [reply include/swift/AST/DiagnosticsParse.def]: Good catch. Done.

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux gpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Review by marcrasi — APPROVED** [review]

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux cpu

**Pull request merged**